fitbenchmarking/parsing/parse_sasview_data.py

```diff
@@ -39,60 +39,68 @@ def __init__(self, fname):
+        tmp_equation = entries['function'].split(',', 1)[0]
+        self._equation = tmp_equation.split('=')[1]
+
+        tmp_starting_values = entries['function'].split(',')[1:]
+        self._starting_values = [[f.split('=')[0], [float(f.split('=')[1])]]
+                                 for f in tmp_starting_values]
+
+        tmp_starting_value_ranges = entries['parameter_ranges'].split(';')
+        tmp_names = (val_range_txt.split('.', 1)[0]
+                     for val_range_txt in tmp_starting_value_ranges)
+        tmp_values = (tmp_range.split('(')[1]
+                      for tmp_range in tmp_starting_value_ranges)
+        tmp_values = (tmp_range.strip(')') for tmp_range in tmp_values)
+        tmp_values = ([float(val) for val in tmp_range.split(',')]
+                      for tmp_range in tmp_values)
+        self._starting_value_ranges = {name: values
+                                       for name, values
+                                       in zip(tmp_names, tmp_values)}
+
+        self.function = None
 
         super(FittingProblem, self).close_file()
 
-    def eval_f(self, x, *param_list):
+    def get_function(self):
         """
-        Function Evaluation Method
-
-        @param x :: x data values
-        @param *param_list :: parameter value(s)
+        Creates list of functions alongside the starting parameters.
+        Functions are saved to the instance so that this is only generated
+        once.
 
-        @ returns :: the y data values evaluated from the model
+        @returns :: function definition list containing the model and its
+                    starting parameter values
         """
-        data = empty_data1D(x)
-        model = load_model((self._equation.split('='))[1])
-
-        param_names = [(param.split('='))[0] for param in self.starting_values.split(',')]
-        if len(param_list) == 1:
-            if isinstance(param_list[0], basestring):
-                exec ("params = dict(" + param_list[0] + ")")
-        else:
-            param_string = ''
-            for name, value in zip(param_names, param_list):
-                param_string += name+'='+str(value)+','
-            param_string = param_string[:-1]
-            exec ("params = dict(" + param_string + ")")
-
-        model_wrapper = Model(model, **params)
-        for range in self.starting_value_ranges.split(';'):
-            exec ('model_wrapper.' + range)
-        func_wrapper = Experiment(data=data, model=model_wrapper)
-
-        return func_wrapper.theory()
+        if self.function is None:
 
-    def get_function(self):
-        """
+            functions = []
+            for i in range(len(self._starting_values[0][1])):
 
-        @returns :: function definition list containing the model and its starting parameter values
-        """
+                param_names = [params[0] for params in self._starting_values]
+
+                def fitFunction(x, *tmp_params):
```

Comment:
And I'm almost certain this function shouldn't be defined in this loop

Reply:
This creates a list of callable functions. I can defer it to a function which returns the function (which would look much like a decorator) but it needs to define a function in this loop somehow

Reply:
Discussed and found that the callable-param pairs are expected to have the same callable so moved. (Also learned that python doesn't like defining functions in loops!)